File: cses.fi/Subordinates.cpp
/*
 * https://cses.fi/problemset/task/1674
 */
#include <bits/stdc++.h>
using namespace std;

vector<vector<int> > graph;
vector<int> dp;

int dfs(int node) {
	int count = 0;
	for (int i = 0; i < graph[node].size(); i++)
		count += 1 + dfs(graph[node][i]);
	return dp[node] = count;
}

int main() {
	int nodes, par, node = 2;
	cin >> nodes;
	graph.resize(nodes + 1);
	for (int i = 0; i < nodes - 1; i++) {
		cin >> par;
		graph[par].push_back(node++);
	}
	for (int i = 0; i <= nodes + 1; i++)
		dp.push_back(-1);
	dfs(1);
	for (int i = 1; i <= nodes; i++)
		cout << dp[i] << ' ';
}



/*
 * Aliter method :-
#include <bits/stdc++.h>
using namespace std;

vector<int> dp;
vector<vector<int> > graph;

int dfs(int node) {
    int ans = 0;
    for (int child : graph[node])
        ans += dfs(child);
    ans += graph[node].size();
    return dp[node] = ans;
}

int main() {
    int n, u;
    cin >> n;
    graph.resize(n + 1);
    for (int i = 2; i <= n; i++) {
        cin >> u;
        graph[u].push_back(i);
    }
    dp.resize(n + 1, 0);
    dfs(1);
    for (int i = 1; i <= n; i++)
        cout << dp[i] << ' ';
    cout << endl;
}
 */

File: cses.fi/Subordinates.java
/*
 * https://cses.fi/problemset/task/1674
 */
import java.io.*;
import java.lang.*;
import java.util.*;


/* 
 * The Java code gives TLE, but the CPP code gives AC!!
 * Solution can be referred from Karthik Arora's video here -> https://www.youtube.com/watch?v=fGznXJ-LTbI&list=PLb3g_Z8nEv1j_BC-fmZWHFe6jmU_zv-8s&t=34s
 */
public class Subordinates {

	private static int[] subtreeNodes;
	private static Map<Integer, List<Integer> > graph = new HashMap<>();

	private static int dfs(int node) {
		if (graph.get(node).isEmpty())
			return subtreeNodes[node] = 0;
		int count = 0;
		for (int neighbor : graph.get(node))
			count += 1 + dfs(neighbor);
		return subtreeNodes[node] = count;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int nodes = Integer.parseInt(br.readLine()), node = 2;
		String[] temp = br.readLine().split(" ");
		for (int i = 1; i <= nodes; i++)
			graph.put(i, new ArrayList<>());
		for (int i = 0; i < nodes - 1; i++) {
			int par = Integer.parseInt(temp[i]);
			graph.computeIfAbsent(par, k -> new ArrayList<>()).add(node++);
		}		
		subtreeNodes = new int[1 + nodes];
		Arrays.fill(subtreeNodes, -1);
		dfs(1);
		for (int i = 1; i <= nodes; i++)
			System.out.print(subtreeNodes[i] + " ");
	}
}
